pkg/csi/service/cns/controller.go

```diff
@@ -0,0 +1,392 @@
+	}
+	node, err := c.nodeMgr.GetNodeByName(req.NodeId)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to find VirtualMachine for node:%q. Error: %v", req.NodeId, err)
+		klog.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	klog.V(4).Infof("Found VirtualMachine for node:%q.", req.NodeId)
+	diskUUID, err := common.AttachVolumeUtil(ctx, c.manager, node, req.VolumeId)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to attach disk: %+q with node: %q err %+v", req.VolumeId, req.NodeId, err)
+		klog.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	publishInfo := make(map[string]string)
+	publishInfo[common.AttributeDiskType] = common.DiskTypeString
+	publishInfo[common.AttributeFirstClassDiskUUID] = common.FormatDiskUUID(diskUUID)
+	resp := &csi.ControllerPublishVolumeResponse{
+		PublishContext: publishInfo,
+	}
+	return resp, nil
+}
+
+// ControllerUnpublishVolume detaches a volume from the Node VM.
+// volume id and node name is retrieved from ControllerUnpublishVolumeRequest
+func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (
+	*csi.ControllerUnpublishVolumeResponse, error) {
+
+	klog.V(4).Infof("ControllerUnpublishVolume: called with args %+v", *req)
+	err := validateVanillaControllerUnpublishVolumeRequest(req)
+	if err != nil {
+		msg := fmt.Sprintf("Validation for UnpublishVolume Request: %+v has failed. Error: %v", *req, err)
+		klog.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	node, err := c.nodeMgr.GetNodeByName(req.NodeId)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to find VirtualMachine for node:%q. Error: %v", req.NodeId, err)
+		klog.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	err = common.DetachVolumeUtil(ctx, c.manager, node, req.VolumeId)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to detach disk: %+q from node: %q err %+v", req.VolumeId, req.NodeId, err)
+		klog.Error(msg)
+		return nil, status.Errorf(codes.Internal, msg)
+	}
+	resp := &csi.ControllerUnpublishVolumeResponse{}
+	return resp, nil
+}
+
+// ValidateVolumeCapabilities returns the capabilities of the volume.
+func (c *controller) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (
+	*csi.ValidateVolumeCapabilitiesResponse, error) {
+
+	klog.V(4).Infof("ControllerGetCapabilities: called with args %+v", *req)
+	volCaps := req.GetVolumeCapabilities()
+	var confirmed *csi.ValidateVolumeCapabilitiesResponse_Confirmed
+	if common.IsValidVolumeCapabilities(volCaps) {
+		confirmed = &csi.ValidateVolumeCapabilitiesResponse_Confirmed{VolumeCapabilities: volCaps}
+	}
+	return &csi.ValidateVolumeCapabilitiesResponse{
+		Confirmed: confirmed,
+	}, nil
+}
+
+func (c *controller) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
+	*csi.ListVolumesResponse, error) {
+
+	klog.V(4).Infof("ListVolumes: called with args %+v", *req)
+	return nil, status.Error(codes.Unimplemented, "")
```

Comment:
You return Unimplemented here, but in the list capabilities, you have RPC_LIST_VOLUMES present. You might want to remove RPC_LIST_VOLUMES since it could potentially screw up other software that will depend on that actually working. As example of such a program, something like Velero.

Reply:
removed `csi.ControllerServiceCapability_RPC_LIST_VOLUMES` from `ControllerGetCapabilities`.